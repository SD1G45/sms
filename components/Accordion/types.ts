import { ReactNode } from "react";

export interface AccordionProps extends React.HTMLAttributes<HTMLDivElement> {
  title: string;
  content: ReactNode;
}
export interface ChevronProps extends React.HTMLAttributes<HTMLButtonElement> {
  active: boolean;
}
