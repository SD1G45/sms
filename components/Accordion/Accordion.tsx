import React, { useState, useRef } from "react";
import {
  AccordionSection,
  StyledButton,
  StyledTitle,
  AccordionContent,
  AccordionText,
  ButtonDiv,
  StyledChevron,
} from "./styles";

import Chevron from "./Chevron";

const Accordion = (props: any) => {
  const [setActive, setActiveState] = useState("");
  const [setHeight, setHeightState] = useState("0px");
  const [setRotate, setRotateState] = useState("accordion__icon ");
  const [rotated, setRotated] = useState(false);

  const content = useRef(null);
  function toggleAccordion() {
    {
      rotated ? setRotated(false) : setRotated(true);
    }
    setActiveState(setActive === "" ? "active" : "");
    setHeightState(
      setActive === "active" ? "0px" : `${props.content.current.scrollHeight}px`
    );
  }
  return (
    <>
      <AccordionSection>
        <StyledButton
          className={`accordion ${setActive}`}
          onClick={toggleAccordion}
        >
          <ButtonDiv>
            <StyledTitle>{props.title}</StyledTitle>
            <StyledChevron
              className={`${setRotate}`}
              rotated={rotated}
              width={10}
              fill={"#777"}
            />
          </ButtonDiv>
        </StyledButton>
        <AccordionContent
          ref={content}
          style={{ maxHeight: `${setHeight}` }}
          className="accordion__content"
        >
          <AccordionText
            dangerouslySetInnerHTML={{ __html: props.content }}
          ></AccordionText>
        </AccordionContent>
      </AccordionSection>
    </>
  );
};

export default Accordion;
