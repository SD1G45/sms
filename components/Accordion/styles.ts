import styled from "styled-components";
import { ChevronProps } from "./types";

export const Accordion = styled.ul`
  list-style: none;
  margin: 0;
  // padding: 0;
  margin-left: 10px;
  padding-left: 10px;
`;

export const ListItem = styled.li`
  border-bottom: 5px solid #ffffff;

  &: not(: first-of-type) {
    border-top: 0;
  }
  padding: 10px;
  background-color: #4881f0;
  margin-left: 5px;
`;

export const Title = styled.h2`
  color: white;
  width: 100%;
  // height: 50%;
  margin: 100;
`;

export const Chevron = styled.button<ChevronProps>`
  & {
    content: "";
    flex-shrink: 0;
    width: 20px;
    height: 18px;
    margin-left: 95%;
    background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16' fill='%23ffffff'%3e%3cpath fill-rule='evenodd' d='M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z'/%3e%3c/svg%3e");
    background-repeat: no-repeat;
    background-size: 100%;
    transition: transform 0.2s ease-in-out;
  }

  &:after {
    content: "";
    flex-shrink: 0;
    width: 18px;
    height: 18px;
    margin-left: auto;
    background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16' fill='%23ffffff'%3e%3cpath fill-rule='evenodd' d='M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z'/%3e%3c/svg%3e");
    background-repeat: no-repeat;
    background-size: 100%;
    transition: transform 0.2s ease-in-out;
  }
  transform: ${(props) => (props.active ? "rotate(180deg)" : "")};
`;

export const ItemContainer = styled.div`
  transition: height 0.2s ease-in-out;
  overflow: hidden;

  // background-color: rgba(72, 240, 229, 0.5);
`;

export const Content = styled.div`
  border-top: 1px solid #ffffff;
  padding: 15px 20px;
`;
