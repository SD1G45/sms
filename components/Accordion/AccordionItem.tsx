import { useEffect, useRef, useState } from "react";
import { getRefValue } from "../../lib/hooks";
import { AccordionProps } from "./types";

import { Chevron, Content, ItemContainer, ListItem, Title } from "./styles";

const AccordionItem = ({
  data,
  isOpen,
  btnOnClick,
}: {
  data: AccordionProps;
  isOpen: boolean;
  btnOnClick: () => void;
}) => {
  const contentRef = useRef<HTMLDivElement>(null);
  const [height, setHeight] = useState(0);

  useEffect(() => {
    if (isOpen) {
      const contentEl = getRefValue(contentRef);

      setHeight(contentEl.scrollHeight);
    } else {
      setHeight(0);
    }
  }, [isOpen]);

  return (
    <ListItem>
      <Title>{data.title}</Title>
      <Chevron onClick={btnOnClick} active={isOpen}></Chevron>
      <ItemContainer style={{ height }}>
        <Content ref={contentRef}>{data.content}</Content>
      </ItemContainer>
    </ListItem>
  );
};

export default AccordionItem;
