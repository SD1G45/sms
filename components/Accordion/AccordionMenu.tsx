import { useState } from "react";
import { AccordionProps } from "./types";
import AccordionItem from "./AccordionItem";
import { Accordion } from "./styles";

const AccordionMenu = ({ items }: { items: Array<AccordionProps> }) => {
  const [currentIdx, setCurrentIdx] = useState(-1);
  const btnOnClick = (idx: number) => {
    setCurrentIdx((currentValue) => (currentValue !== idx ? idx : -1));
  };

  return (
    <Accordion>
      {items.map((item, idx) => (
        <AccordionItem
          key={idx}
          data={item}
          isOpen={idx === currentIdx}
          btnOnClick={() => btnOnClick(idx)}
        />
      ))}
    </Accordion>
  );
};

export default AccordionMenu;
