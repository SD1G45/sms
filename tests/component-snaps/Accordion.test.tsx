import React from "react";
import renderer from "react-test-renderer";
import "jest-styled-components";
import Accordion from "../../components/Accordion";

const items = [
  {
    title: "How to Create a Campaign",
    content: (
      <div>
        {/* <strong>Body Header</strong> Body text <u>underline tag</u>,{" "} */}
        {/* <i>italic</i> */}
        Click <b>Create New</b> on the menu under the Campaigns
        <div>
          <ul>
            <br></br>
            <li>Set your campaign title</li>
            <li>Set a message you would like to send</li>
            <li>Set the coupons you'd like to send</li>
            <li>Set the customer list.d</li>
            <br></br>
          </ul>
        </div>
        Click the blue <b>Create Campaign</b> button to continue.
      </div>
    ),
  },
]

describe("<Accordion/>", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Accordion items={items}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});