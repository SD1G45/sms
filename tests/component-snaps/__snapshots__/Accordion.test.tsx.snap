// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion/> renders correctly 1`] = `
.c0 {
  list-style: none;
  margin: 0;
  margin-left: 10px;
  padding-left: 10px;
}

.c1 {
  border-bottom: 5px solid #ffffff;
  padding: 10px;
  background-color: #4881f0;
  margin-left: 5px;
}

.c1:not(:first-of-type) {
  border-top: 0;
}

.c2 {
  color: white;
  width: 100%;
  margin: 100;
}

.c3 {
  content: "";
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 20px;
  height: 18px;
  margin-left: 95%;
  background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16' fill='%23212529'%3e%3cpath fill-rule='evenodd' d='M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z'/%3e%3c/svg%3e");
  background-repeat: no-repeat;
  background-size: 100%;
  -webkit-transition: -webkit-transform 0.2s ease-in-out;
  -webkit-transition: transform 0.2s ease-in-out;
  transition: transform 0.2s ease-in-out;
}

.c3:after {
  content: "";
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 18px;
  height: 18px;
  margin-left: auto;
  background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16' fill='%23212529'%3e%3cpath fill-rule='evenodd' d='M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z'/%3e%3c/svg%3e");
  background-repeat: no-repeat;
  background-size: 100%;
  -webkit-transition: -webkit-transform 0.2s ease-in-out;
  -webkit-transition: transform 0.2s ease-in-out;
  transition: transform 0.2s ease-in-out;
}

.c4 {
  -webkit-transition: height 0.2s ease-in-out;
  transition: height 0.2s ease-in-out;
  overflow: hidden;
}

.c5 {
  border-top: 1px solid #ffffff;
  padding: 15px 20px;
}

<ul
  className="c0"
>
  <li
    className="c1"
  >
    <h2
      className="c2"
    >
      How to Create a Campaign
    </h2>
    <button
      className="c3"
      onClick={[Function]}
    />
    <div
      className="c4"
      style={
        Object {
          "height": 0,
        }
      }
    >
      <div
        className="c5"
      >
        <div>
          Click 
          <b>
            Create New
          </b>
           on the menu under the Campaigns
          <div>
            <ul>
              <br />
              <li>
                Set your campaign title
              </li>
              <li>
                Set a message you would like to send
              </li>
              <li>
                Set the coupons you'd like to send
              </li>
              <li>
                Set the customer list.d
              </li>
              <br />
            </ul>
          </div>
          Click the blue 
          <b>
            Create Campaign
          </b>
           button to continue.
        </div>
      </div>
    </div>
  </li>
</ul>
`;
